File: 00645. Set Mismatch/00645.go
// https://leetcode.com/problems/set-mismatch

func findErrorNums(nums []int) []int {
    sort.Ints(nums)

    size, dup, lost := len(nums), 0, 0
    if nums[0] != 1 {
        lost = 1
    }
    for i := 1; i < size && (dup == 0 || lost == 0); i++ {
        if dup == 0 && nums[i-1] == nums[i] {
            dup = nums[i]
        }
        if lost == 0 && nums[i-1]+1 < nums[i] {
            lost = nums[i-1]+1
        }
    }
    if lost == 0 {
        lost = size
    }

    return []int{dup, lost}
}

File: LeetCode/00645. Set Mismatch/00645.c
// https://leetcode.com/problems/set-mismatch

int compare(const void *a, const void *b) {
    return *(int *)a - *(int *)b;
}

int * findErrorNums(int *nums, int numsSize, int *returnSize) {
    qsort(nums, numsSize, sizeof(int), compare);

    *returnSize = 2;
    int *ans = (int *) malloc(sizeof(int) * 2);
    ans[0] = 0;
    ans[1] = (nums[0] == 1) ? 0 : 1;

    for (int i = 1, tmp = 1; i < numsSize; ++i) {
        if (nums[i-1] == nums[i]) {
            ans[0] = nums[i];
        } else if (nums[i-1]+1 < nums[i]) {
            ans[1] = nums[i-1]+1;
        }
    }

    if (ans[1] == 0) {
        ans[1] = numsSize;
    }
    return ans;
}
